[WIP needs review from developer] Updating OpenAPI tools info

sishida opened the pull request:
Signed-off-by: Sarah Ishida <sishida@us.ibm.com>

Updating OpenAPI tools documentation for 0.6.0.

sishida commented:
Questions:
1️⃣Do we need to update the Eclipse version of this page with similar information? https://www.eclipse.org/codewind/open-api-tools-for-eclipse.html
2️⃣In the warning message that I reviewed in https://github.com/eclipse/codewind-openapi-vscode/pull/66, I don't think we mention that comments are removed. Especially if messages are the only things removed during a merging of pom.xml files, could we mention that in the warning message itself?
3️⃣Should we provide more instructions on how to customize application logic (step 8)?
4️⃣Should we explain how the Spring project instructions relate to the Maven instructions? For example, would you do the Spring instructions after a certain step of the Maven instructions?

keithchong commented on docs/_documentations/codewind-openapi-vscode.md:
```diff
@@ -22,6 +22,21 @@ The Codewind OpenAPI Tools for VS Code provides commands that invoke the OpenAPI
 3. This extension pulls the [OpenAPI Generator CLI Docker Image](https://github.com/OpenAPITools/openapi-generator#16---docker) and runs the OpenAPI Generator in a Docker container. Install Docker if necessary.
 4. Install the Codewind OpenAPI Tools from the [VS Code Marketplace](https://marketplace.visualstudio.com/items?itemName=IBM.codewind-openapi-tools) or by searching for `Codewind OpenAPI` in the [VS Code Extensions view](https://code.visualstudio.com/docs/editor/extension-gallery#_browse-for-extensions).
 
+## Generating and building code in an existing Java Maven project
```
I'd like this section to be for the Open API Tools for VS Code (NOT, the tools for eclipse, because the steps and workaround are different)

sishida replied:
@keithchong This PR update is for a document that only applies to VS Code tools. Eclipse would need to be a separate update. Is a title like "Generating and building code and clients for OpenAPI tools" suitable? Or is there a term that would cover both "code" and "clients" together?

keithchong replied:
The title can be improved.  'Code' is too general and we need to mention 'server'.  Change the title from `Generating and building code and clients for OpenAPI tools` to something like:
`Generating and building client and server stubs using the OpenAPI tools`

keithchong commented on docs/_documentations/codewind-openapi-vscode.md:
```diff
@@ -22,6 +22,21 @@ The Codewind OpenAPI Tools for VS Code provides commands that invoke the OpenAPI
 4. Install the Codewind OpenAPI Tools from the [VS Code Marketplace](https://marketplace.visualstudio.com/items?itemName=IBM.codewind-openapi-tools) or by searching for `Codewind OpenAPI` in the [VS Code Extensions view](https://code.visualstudio.com/docs/editor/extension-gallery#_browse-for-extensions).
 
+## Generating and building code in an existing Java Maven project
+1. Right-click your Codewind project and launch the **Generate Server from OpenAPI Definition** wizard.
```
Right-click your Codewind project from the Codewind view (in VS Code) and select either the **Generate Server from OpenAPI Definition** menu or **Generate Client from OpenAPI Definition**action.  You will be prompted to provide options to the Open API generator.

keithchong commented on docs/_documentations/codewind-openapi-vscode.md:
```diff
@@ -22,6 +22,21 @@ The Codewind OpenAPI Tools for VS Code provides commands that invoke the OpenAPI
 
+## Generating and building code in an existing Java Maven project
+1. Right-click your Codewind project and launch the **Generate Server from OpenAPI Definition** wizard.
+2. Select your OpenAPI document `.json` file.
```
2. Select your OpenAPI document file.
They have to select a valid Open API 3.0 YAML or JSON document that is in the project

keithchong commented on docs/_documentations/codewind-openapi-vscode.md:
```diff
@@ -22,6 +22,21 @@ The Codewind OpenAPI Tools for VS Code provides commands that invoke the OpenAPI
+## Generating and building code in an existing Java Maven project
+1. Right-click your Codewind project and launch the **Generate Server from OpenAPI Definition** wizard.
+2. Select your OpenAPI document `.json` file.
+3. Select the `jaxrs-spec` server generator type.
```
3. Select the client or server generator type.

keithchong commented on docs/_documentations/codewind-openapi-vscode.md:
```diff
@@ -22,6 +22,21 @@ The Codewind OpenAPI Tools for VS Code provides commands that invoke the OpenAPI
+1. Right-click your Codewind project and launch the **Generate Server from OpenAPI Definition** wizard.
+2. Select your OpenAPI document `.json` file.
+3. Select the `jaxrs-spec` server generator type.
+4. When you're prompted to select the output folder, keep the default selection. The project root, which is the parent of the source folder, is selected by default.
```
This is good.   Technically, you can generate into a child folder of a maven project, but the user will have to refactor the whole project to be multi-module.  This is beyond the scope of this enhancement.  We want to leave this option available in case they do want to do the codegen and refactor it themselves.

keithchong commented on docs/_documentations/codewind-openapi-vscode.md:
```diff
@@ -22,6 +22,21 @@ The Codewind OpenAPI Tools for VS Code provides commands that invoke the OpenAPI
+2. Select your OpenAPI document `.json` file.
+3. Select the `jaxrs-spec` server generator type.
+4. When you're prompted to select the output folder, keep the default selection. The project root, which is the parent of the source folder, is selected by default.
+5. Click **Select As Output Folder** to merge the generated `.pom` file with the existing `.pom` file. The merge attempts to include all necessary dependencies for the project to build properly.
```
5. Click **Select As Output Folder**.  
Move the rest to after step 6.

sishida replied:
@keithchong It's better to include the results of an action as part of the step that includes the action. When you click **Select As Output Folder**, is that action what merges the `.pom` files? If so, it's probably best to leave these statements together.

keithchong replied:
@sishida , that is the last step.  (Unlike in Eclipse, you can click on the Finish button of the wizard). Then the warning message appears right after.  If the user cancels the warning, then _no_ code generation and _no_ merging takes place.

sishida replied:
OK, I think I understand better what's going on. I will edit and add a new commit.

keithchong commented on docs/_documentations/codewind-openapi-vscode.md:
```diff
@@ -22,6 +22,21 @@ The Codewind OpenAPI Tools for VS Code provides commands that invoke the OpenAPI
+3. Select the `jaxrs-spec` server generator type.
+4. When you're prompted to select the output folder, keep the default selection. The project root, which is the parent of the source folder, is selected by default.
+5. Click **Select As Output Folder** to merge the generated `.pom` file with the existing `.pom` file. The merge attempts to include all necessary dependencies for the project to build properly.
+6. When prompted with the warning message, note that the `pom.xml` file is backed up. However, XML comments, such as `<!-- example comment -->`, are removed. Click **Yes** to proceed. The files build.
```
6. At this step, the currently supported generator options have been collected. Code generation is ready to begin until after a warning message to proceed is confirmed.  The existing `pom.xml` file is backed up first as `pom-backup.xml` or variation.  The new `pom.xml` will be a merged version of the original POM and the one that is generated.  The merge attempts to include all necessary dependencies for the project to build properly.  Due to limitations in the merge process, XML comments such as `<!-- example comment -->`, from the original POM are not preserved.   When prompted with the warning message, click **Yes** to proceed with code generation.  For any Codewind project, if autobuild is turned on, the project will build automatically.

keithchong commented on docs/_documentations/codewind-openapi-vscode.md:
```diff
@@ -22,6 +22,21 @@ The Codewind OpenAPI Tools for VS Code provides commands that invoke the OpenAPI
+4. When you're prompted to select the output folder, keep the default selection. The project root, which is the parent of the source folder, is selected by default.
+5. Click **Select As Output Folder** to merge the generated `.pom` file with the existing `.pom` file. The merge attempts to include all necessary dependencies for the project to build properly.
+6. When prompted with the warning message, note that the `pom.xml` file is backed up. However, XML comments, such as `<!-- example comment -->`, are removed. Click **Yes** to proceed. The files build.
+7. See that your `src` folder shows the newly generated code complete with a `gen` package. Your project state returns to the `[Running]` state. The `src/main/java` and `src/test/java` are not customizable.
```
7. See that your `src` folder shows the newly generated code. Your Codewind project state returns to the `[Running]` state. The target source folders `src/main/java` and `src/test/java` are currently not customizable.

(Removed: `complete with a `gen` package` because the gen folder is specific to the generator that is chosen)

keithchong commented on docs/_documentations/codewind-openapi-vscode.md:
```diff
@@ -22,6 +22,21 @@ The Codewind OpenAPI Tools for VS Code provides commands that invoke the OpenAPI
+8. Now, you can customize the application logic.
+
+## Generating and building code in an existing Java Spring project
+When you generate into a Spring project, don't override the implicitly or explicitly configured main class that comes with the project.
```
When you generate a Spring server stub into a Spring project, the Open API tools for VS Code does not override the main class that is already implicitly or explicitly configured in the project.  There are manual steps to expose the OpenAPI endpoints:

keithchong commented on docs/_documentations/codewind-openapi-vscode.md:
```diff
@@ -22,6 +22,21 @@ The Codewind OpenAPI Tools for VS Code provides commands that invoke the OpenAPI
+
+## Generating and building code in an existing Java Spring project
+When you generate into a Spring project, don't override the implicitly or explicitly configured main class that comes with the project.
+1. To use the generated class, comment out the main class text in the project `.java` file.
```
1. If you want to use the generated class, uncomment the main method in the Java class `OpenAPI2SpringBoot.java` and explicitly configure the project to use this main class.

keithchong commented on docs/_documentations/codewind-openapi-vscode.md:
```diff
@@ -22,6 +22,21 @@ The Codewind OpenAPI Tools for VS Code provides commands that invoke the OpenAPI
+## Generating and building code in an existing Java Spring project
+When you generate into a Spring project, don't override the implicitly or explicitly configured main class that comes with the project.
+1. To use the generated class, comment out the main class text in the project `.java` file.
+2. To use the current application, copy the packages at the component scan annotation and move them to the `SBApplication.java` file. Then, launch the project.
```
2. To continue to use the currently configured main class, copy the base packages listed in the component scan annotation `@ComponentScan` from `OpenAPI2SpringBoot.java` and move them to the currently configured main class. Then, launch the project.   For example, if you are using the Codewind Spring Boot project, the implicit main class is the `SBApplication.java` file.

keithchong commented on docs/_documentations/troubleshooting.md:
````diff
@@ -539,7 +539,7 @@ The build is successful even though the validator reports this issue.
 ```
 
-The following work-around applies to both VS Code and Eclipse. The spring server generator creates invalid source in the `OpenAPI2SpringBoot` class. Simply implement the methods from the interface and save the file. Also add the configuration element to the `pom.xml` file, like this:
+The following workaround applies to Eclipse. The Spring server generator creates invalid source in the `OpenAPI2SpringBoot` class. Simply implement the methods from the interface and save the file. Also add the configuration element to the `pom.xml` file:
````
This change looks good.

sishida commented:
@keithchong I've added more commits and am ready for the next round of review! Please let me know if you'd like more edits or if we're good to go. Thanks!

keithchong commented:
> Questions:
> 1️⃣Do we need to update the Eclipse version of this page with similar information? https://www.eclipse.org/codewind/open-api-tools-for-eclipse.html
Yes good point.  Step 6 and 7 should be appended to step 5&6 in the Eclipse help.
Do not add the section "Generating and building code in an existing Java Spring project" to the Eclipse help because we don't have that same support for Eclipse.
> 2️⃣In the warning message that I reviewed in [eclipse/codewind-openapi-vscode#66](https://github.com/eclipse/codewind-openapi-vscode/pull/66), I don't think we mention that comments are removed. Especially if messages are the only things removed during a merging of pom.xml files, could we mention that in the warning message itself?
I got feedback that the warning message was too long and included too much detail.  The more important points are included (POM merge and backup)
> 3️⃣Should we provide more instructions on how to customize application logic (step 8)?
No. It's specific to the application.
> 4️⃣Should we explain how the Spring project instructions relate to the Maven instructions? For example, would you do the Spring instructions after a certain step of the Maven instructions?
The Spring instructions should be done right after the code generation is completed.

sishida commented:
@keithchong In my latest commit...
- I have updated the heading title as you suggested.
- I have also added the section to the Eclipse tools page. However, I wasn't sure how combining steps would make the new section more relevant to Eclipse? Could you please comment on the lines that you'd like changed on the Eclipse page (now added to this PR)? Thank you!

keithchong commented on docs/_documentations/codewind-openapi-eclipse.md:
```diff
@@ -18,6 +18,16 @@ The Codewind OpenAPI Tools for Eclipse includes wizards that invoke the OpenAPI
+6. When prompted with the warning message, click **Yes** to merge the generated `.pom` file with the existing `.pom` file. The existing `pom.xml` file is backed up as `pom-backup.xml` or as a variation of this name. The new `pom.xml` file is a merged version of the original `.pom` file and the `.pom` file that is generated. The merge attempts to include all necessary dependencies for the project  ...
+7. See that your `src` folder shows the newly generated code. Depending on the generator type that you chose, a `gen` package is created. Your Codewind project state returns to the `[Running]` state. The target source folders, `src/main/java` and `src/test/java`, are not customizable.
+8. Now, you can customize the application logic.
+
```
@sishida , the Eclipse steps are different, so we can't use the steps from VS Code here.  Steps 6 to 8 above however are common enough to repeat in the Eclipse help.  Those steos should be appended to step 5 or 6 under the Running commands section below.

S

keithchong replied:
Under the `Running commands` section, step 4, `Select an action to generate the client or server stubs or HTML documentation` essentially is the start of this _new_ section entitled `Generating and building client and server stubs by using the OpenAPI tools`

keithchong replied:
For example, something like this:
4. Select an action to generate the client or server stubs or HTML documentation.
5. If client or server stub is selected,
6. Enter the requested information that is displayed in the wizard.
- For output folder, keep the default selection. The project root, which is the parent of the source folder, is selected by default.
7. Click Finish
8. When prompted with the warning message, click **Yes** to merge the generated `.pom` file with the existing `.pom` file. The existing `pom.xml` file is backed up as `pom-backup.xml` or as a variation of this name. The new `pom.xml` file is a merged version of the original `.pom` file and the `.pom` file that is generated. The merge attempts to include all necessary dependencies for the project to build properly. Because of limitations in the merge process, XML comments, such as `<!-- example comment -->`, from the original `.pom` file are removed. For any Codewind project, if autobuild is turned on, the project builds automatically.
9. See that your `src` folder shows the newly generated code. Depending on the generator type that you chose, a `gen` package is created. Your Codewind project state returns to the `[Running]` state. The target source folders, `src/main/java` and `src/test/java`, are not customizable.
10. Now, you can customize the application logic.
11. After generation, edit the .openapi-generator-ignore file to ensure that subsequent code generation does not overwrite custom code.


After generation, edit the .openapi-generator-ignore file to ensure that subsequent code generation does not overwrite custom code.

sishida commented:
@keithchong I have edited the Eclipse file according to your latest comments. Thank you for your explanations!

Because we combined the "Running commands" section with the new "Generating and building client and server stubs by using the OpenAPI tools" section, I retitled the whole section to "Running commands, generating, and building client and server stubs by using the OpenAPI tools." Please let me know if you'd like a different one. Thanks!

keithchong commented on docs/_documentations/codewind-openapi-eclipse.md:
```diff
@@ -18,13 +18,17 @@ The Codewind OpenAPI Tools for Eclipse includes wizards that invoke the OpenAPI
 3. Install the [Codewind OpenAPI Tools from the Eclipse Marketplace](https://marketplace.eclipse.org/content/codewind).
 
-## Running commands
+## Running commands, generating, and building client and server stubs by using the OpenAPI tools
```
Step 4 mentions HTML docs too.  I'd remove `Running commands`.   How about:

Generating HTML, and client and server stubs by using the OpenAPI wizard

sishida commented:
@keithchong I have updated the heading!
